jormungandr/src/explorer/graphql/mod.rs

```diff
@@ -131,99 +152,142 @@ impl Branch {
+            |after, before, first, last| async move {
+                let boundaries = if !vote_plans.is_empty() {
+                    PaginationInterval::Inclusive(InclusivePaginationInterval {
+                        lower_bound: 0u32,
+                        upper_bound: vote_plans
+                            .len()
+                            .checked_sub(1)
+                            .unwrap()
+                            .try_into()
+                            .expect("tried to paginate more than 2^32 elements"),
                     })
-                    .collect::<Vec<(VotePlanStatus, u32)>>()
-            }
-        })
+                } else {
+                    PaginationInterval::Empty
+                };
+
+                let pagination_arguments = ValidatedPaginationArguments {
+                    first,
+                    last,
+                    before: before.map(u32::try_from).transpose()?,
+                    after: after.map(u32::try_from).transpose()?,
+                };
+
+                let (range, page_meta) = compute_interval(boundaries, pagination_arguments)?;
+                let mut connection = Connection::with_additional_fields(
+                    page_meta.has_previous_page,
+                    page_meta.has_next_page,
+                    ConnectionFields {
+                        total_count: page_meta.total_count,
+                    },
+                );
+
+                let edges = match range {
+                    PaginationInterval::Empty => vec![],
+                    PaginationInterval::Inclusive(range) => {
+                        let from = range.lower_bound;
+                        let to = range.upper_bound;
+
+                        (from..=to)
+                            .map(|i: u32| {
+                                let (_pool_id, vote_plan_data) =
+                                    &vote_plans[usize::try_from(i).unwrap()];
+                                (
+                                    VotePlanStatus::vote_plan_from_data(Arc::clone(vote_plan_data)),
+                                    i,
+                                )
+                            })
+                            .collect::<Vec<(VotePlanStatus, u32)>>()
+                    }
+                };
+
+                connection.append(
+                    edges
+                        .iter()
+                        .map(|(vps, cursor)| Edge::new(IndexCursor::from(*cursor), vps.clone())),
+                );
+
+                Ok(connection)
+            },
+        )
+        .await
     }
 
     pub async fn all_stake_pools(
         &self,
         first: Option<i32>,
         last: Option<i32>,
-        before: Option<IndexCursor>,
-        after: Option<IndexCursor>,
-    ) -> FieldResult<PoolConnection> {
+        before: Option<String>,
+        after: Option<String>,
+    ) -> FieldResult<Connection<IndexCursor, Pool, ConnectionFields<PoolCount>, EmptyFields>> {
```

Comment:
So the `before`/`after` parameters are not typed now because of the proc macro, but there is still an `IndexCursor` type in the return?
Not going into details, I assume this is how async-graphql restricts these pagination APIs.

Reply:
The `query` macro expects strings, that's why I changed the user-facing api too, because it will perform validation twice, otherwise. 

I'm not sure why couldn't the macro accept anything that implements `CursorType`, to be honest, as it lets you return it.

Now I'm thinking that it is a bit weird to have the input cursor types be different than the type in the `Connection`, so I think I'll change it so they are both `IndexCursor`. (which is not against the spec, anyway, and just requires unwrapping a newtype before passing to the macro)